pkg/testing/linearizer/linearizer.go

```diff
@@ -0,0 +1,252 @@
+package linearizer
+
+import (
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"testing"
+)
+
+/*
+   A controller has this sequence: Get, Create, and Update
+   Two instances of the controller: A, and B
+   The two instances run on their own goroutine, and we try to exercise
+   all possible combinations:
+     -- A/Start -> A/Get   -> A/Create -> A/Update -> B/Start  -> B/Get    -> B/Create -> B/Update
+     -- A/Start -> B/Start -> A/Get    -> A/Create -> A/Update -> B/Get    -> B/Create -> B/Update
+     -- A/Start -> A/Get   -> B/Start  -> A/Create -> A/Update -> B/Get    -> B/Create -> B/Update
+     -- A/Start -> A/Get   -> A/Create -> B/Start  -> A/Update -> B/Get    -> B/Create -> B/Update
+     -- A/Start -> B/Start -> B/Get    -> A/Get    -> A/Create -> A/Update -> B/Create -> B/Update
+     -- A/Start -> B/Start -> A/Get    -> B/Get    -> A/Create -> A/Update -> B/Create -> B/Update
+     -- A/Start -> A/Get   -> B/Start  -> B/Get    -> A/Create -> A/Update -> B/Create -> B/Update
+     -- A/Start -> B/Start -> A/Get    -> A/Create -> B/Get    -> A/Update -> B/Create -> B/Update
+     -- (more)
+*/
+
+// Actor represents a single goroutine
+type Actor interface {
+	// Must be invoked to wait as soon as an actor enters a goroutine
+	Start() <-chan struct{}
+	// Each API call is prepended with this so the dispatcher
+	// can linearize invocations
+	Next(op string) <-chan struct{}
+	// Must be invoked as soon as an actor exits a goroutine
+	Exit()
+}
+
+type Reporter interface {
+	Report()
+}
+
+// Dispatcher helps with serializing the actors
+type Dispatcher interface {
+	Reporter
+
+	Register(actorID int) Actor
+
+	Wait(op string) <-chan struct{}
+
+	// Should be called as soon as all actors have
+	// exited from their goroutines
+	PrepareNext() bool
+
+	UseT(t *testing.T)
+}
+
+func NewDispatcher(t *testing.T, actors int, notify func(id int, op string)) Dispatcher {
+	d := &dispatcher{
+		t:        t,
+		count:    actors,
+		selector: &exhaustive{t: t},
+		notifyFn: notify,
+	}
+	d.active.Store(int64(d.count))
+	d.actors = map[string]*actor{}
+	for i := 0; i < actors; i++ {
+		d.actors[label(i)] = &actor{id: i, d: d}
+	}
+	return d
+}
+
+func label(actorID int) string { return fmt.Sprintf("actor-%d", actorID) }
+
+type actor struct {
+	id int
+	d  *dispatcher
+}
+
+func (a actor) Start() <-chan struct{} {
```

Comment:
it's so that we can linearize the startup logic as well, the test here doesn't instrument the lister logic and a controller can start with a get from the lister.
A lister reacts to the state of the storage, so as long as we linearize all the API invocations we are good, this also means that the startup logic must me linearized as well.